File: 1052.爱生气的书店老板/solution.cpp
class Solution {
public:
    int maxSatisfied(vector<int>& customers, vector<int>& grumpy, int X) {
        int sz=customers.size();
        int all=0;
        for (int i=0;i<sz;i++) {
            if (grumpy.at(i)==0) {
                all+=customers.at(i);
            }
        }
        int ans=all;
        int win=0;
        for (int i=0;i<X-1;i++) {
            if (grumpy.at(i)==1) {
                win+=customers.at(i);
            }
        }
        for (int i=X-1;i<sz;i++) {
            if (grumpy.at(i)==1) {
                win+=customers.at(i);
            }
            if (i-X>=0 && grumpy.at(i-X)==1) {
                win-=customers.at(i-X);
            }
            ans=max(ans,win+all);
        }
        return ans;
    }
};

File: 1052.爱生气的书店老板/solution.go
func maxSatisfied(customers []int, grumpy []int, X int) int {
	var all int
	sz := len(customers)
	for i := 0; i < sz; i++ {
		if grumpy[i] == 0 {
			all += customers[i]
		}
	}
	var ans = all
	var win int
	for i := 0; i < X-1; i++ {
		if grumpy[i] == 1 {
			win += customers[i]
		}
	}
	for i := X - 1; i < sz; i++ {
		if grumpy[i] == 1 {
			win += customers[i]
		}
		if i-X >= 0 && grumpy[i-X] == 1 {
			win -= customers[i-X]
		}
		ans = max(ans, win+all)
	}
	return ans
}
func max(a, b int) int {
	if a >= b {
		return a
	}
	return b
}
